### ISSUE TITLE
Build For Apple Silicon (new Github runners)

### ISSUE THREAD

**Issue by gal064** [pbek/QOwnNotes#2912]
Github [runners for M1](https://github.blog/2023-10-02-introducing-the-new-apple-silicon-powered-m1-macos-larger-runner-for-github-actions/) are now available and I'm not familiar with QT, but it looks like you could even [build an M1 version on intel silicon](https://www.qt.io/blog/qt-on-apple-silicon). 

Would it be possible to have an M1 build?

**Comment by pbek** [comment]
The build process for macos would need to work on Qt6 to do that I guess.
And I don't know if you could get two different architectures in one disk image with the Qt tooling.
But I also have no idea if the update script would be able to handle that.

And I'm lacking M1 hardware to be able to test the result 😁

**Comment by pbek** [comment]
> I'm happy to help however I can. Unfortunately know nothing about QT to actually prepare a full PR, but if you want me to run a certain GitHub action or script, or test the result happy to do so.

Thank you!
If anyone wants to pick this up, this is where the current release build process is started (uses Qt5):
https://github.com/pbek/QOwnNotes/blob/952a3684e9840995673545b1509e7b1f286faadc/.github/workflows/build-release.yml#L331-L395
Most of the actual process is abstracted to https://github.com/pbek/QOwnNotes/blob/main/build-systems/github/macos/create-dmg.sh.

In case Qt6 is needed, this is where I experimentally build the macOS build with Qt6:
https://github.com/pbek/QOwnNotes/blob/952a3684e9840995673545b1509e7b1f286faadc/.github/workflows/build-release-qt6.yml#L355-L410


> BTW QTOwnNotes do sometimes freeze for a few seconds on M1. I don't know if it's just a bug, or related to not being built for apple Silicon, but in the past I've seen weird behaviors like this happen on intel apps run on M1

Any debug logs?

**Comment by pbek** [comment]
Building should maybe even work with Qt5 and qmake says: https://doc-snapshots.qt.io/qt5-5.15/macos.html#architectures

@gal064, so in theory you could try to fork the QOwnNotes repository and add ` QMAKE_APPLE_DEVICE_ARCHS="x86_64 arm64"` to this line:

https://github.com/pbek/QOwnNotes/blob/952a3684e9840995673545b1509e7b1f286faadc/.github/workflows/build-release.yml#L377

If you enabled GitHub actions and push your changes to the `release` branch, that should trigger building a release for you.

**Comment by gal064** [comment]
Tried to run it but I'm getting:
`"Trying to compile Botan configured as x86_64 with non-x86_64 compiler."`

Also tried to comment out the x86 config in `botan.pri` and received the following:
`ld: symbol(s) not found for architecture arm64`

I'm not sure if that error is unrelated or making changes to the pri file created that error. I was pretty much guessing

**Comment by gal064** [comment]
Yep. I have no idea how to push forward here, so if you or anyone else want to pick this up would be greatly appreciated.

Anyway I'll submit a bug on the small freezes when it happens next

**Comment by pbek** [comment]
I tried just adding `QMAKE_APPLE_DEVICE_ARCHS = x86_64 arm64` to the project file in https://github.com/pbek/QOwnNotes/commit/fdd5e059473a27f8e26a802fdf624b94f522017e, but the build on macOS failed, without telling why. 🤔
https://github.com/pbek/QOwnNotes/actions/runs/7685776723/job/20943900082

**Comment by spoelstraethan** [comment]
I wonder if any of the notes here on cross compilation might help/work?

https://github.com/qbittorrent/qBittorrent/wiki/Compilation:-macOS-(x86_64,-arm64,-cross-compilation)#qt

**Comment by gal064** [comment]
Looking at the logs, can the failure might be due to this:
```
:17:
libraries/botan/botan.h:293:6: error: "Trying to compile Botan configured as x86_64 with non-x86_64 compiler."
    #error "Trying to compile Botan configured as x86_64 with non-x86_64 compiler."
     ^
```

**Comment by gal064** [comment]
I ran into the same issue but wasn't sure how to set it up

**Comment by pbek** [comment]
Ah, good find! Hm... 🤔

**Comment by KwisatzJim** [comment]
I have a MBPro M3 Pro and was able to build for Apple Silicon.  I did the fallowing:
git clone https://github.com/pbek/QOwnNotes.git -b release --depth=1
cd QOwnNotes
git submodule update --init

brew install qt-creator

brew install qt
You can add Homebrew's Qt to QtCreator's "Qt Versions" in:
  Preferences > Qt Versions > Link with Qt...
pressing "Choose..." and selecting as the Qt installation path:
  /opt/homebrew

brew install qt@5
fallow procedure to add qt5 to QT Creator:
You can add Homebrew's Qt to QtCreator's "Qt Versions" in:
  Preferences > Qt Versions > Link with Qt...
pressing "Choose..." and selecting as the Qt installation path:
  /opt/homebrew/opt/qt@5

open QT Creator
open the project file src/QOwnNotes.pro

then I think there was a 'configure project' button but it's not there now and I don't recall exactly, sorry. But that did it's thing for a while. I've never used QT Creator before so going by memory.

then I built it and ran it and it worked, building an Apple Silicon version

**Comment by KwisatzJim** [comment]
hmmm... ran into a hitch, I copied the .app to my Mac Mini M2 and it crashed on start until I installed qt@5

**Comment by pbek** [comment]
You could try to use the package manager nix to install QOwnNotes on the mac. 🤔

**Comment by KwisatzJim** [comment]
> You could try to use the package manager nix to install QOwnNotes on the mac. 🤔

ok, not bad idea, but the nix version is 24.4.0 and current release is 24.4.2.  But it is Apple Silicon which is nice.

**Comment by pbek** [comment]
Will be surely in the build pipeline. QOwnNotes is at 24.4.2 in nix.

**Comment by pbek** [comment]
I'd like to know if it worked in case you try 😉

**Comment by sfatula** [comment]
It does not build even with nix 24.11, and nix lists it as available for aarch64 darwin, it is not. Same error as I reported elsewhere here:

       error: Package ‘xvfb-run-1+g87f6705’ in /nix/store/g7wwck770vsq4df6xli5i4gcyhgqp13q-source/pkgs/tools/misc/xvfb-run/default.nix:59 is not available on the requested hostPlatform:

**Comment by pbek** [comment]
Hm, that's just to run the (graphical) application to generate the shell completion. 🤔

https://github.com/NixOS/nixpkgs/blob/d0797a04b81caeae77bcff10a9dde78bc17f5661/pkgs/applications/office/qownnotes/default.nix#L56-L61

**Comment by sfatula** [comment]
But nix errors out so stuff does not build. What's the solution?

**Comment by pbek** [comment]
Can you please try building the app without the whole block from https://github.com/NixOS/nixpkgs/blob/d0797a04b81caeae77bcff10a9dde78bc17f5661/pkgs/applications/office/qownnotes/default.nix#L56-L61?

**Comment by pbek** [comment]
You can build it like I do here:
https://github.com/pbek/nixcfg/blob/24a215df4b3b114f388f73e38f83fd5ddc3cfee9/apps/qownnotes/justfile#L2

**Comment by sfatula** [comment]
Ok, so, this is mostly above my pay grade, I don't mess with nix. Still I was able to remove that block, and, I also removed the other 2 lines in default.nix that had xvfb in them, and, I got it to build but it doesn't make the .app file. So, the error that comes out from nix is:

       > mv: cannot stat '/nix/store/nciprrckazcz8fdqydn0l5w68v4794pn-qownnotes-24.11.1/bin/QOwnNotes.app': No such file or directory

I found the binary (not .app), executed it and it ran. I'd prefer nix work though so I don't have to manually build things, it's not something I am terribly familiar with.

**Comment by pbek** [comment]
> I'd prefer nix work though so I don't have to manually build things, it's not something I am terribly familiar with.

The goal is to create a working package for macOS, and you are the only one I know with Apple hardware. The initial package for NixOS was not created by me. Your issue belongs to https://github.com/NixOS/nixpkgs/issues.

This is where the app file gets moved around in the end:

https://github.com/NixOS/nixpkgs/blob/22c3f2cf41a0e70184334a958e6b124fb0ce3e01/pkgs/applications/office/qownnotes/default.nix#L67-L72

So where did you find the binary that worked? Can you adapt the paths in above postInstall script?

**Comment by sfatula** [comment]
I see, so, I am 100% certain I did not do it the right way or even useable way. I don't care if nix works per se, what I am looking for is simply a native Mac build (on Apple Silicon not Intel) of qownnotes by any method except building my own. All the downloadable ones are Intel. 

While the build works, the main toolbar is messed up in that half the icons are missing and replaced with words (just a cursory review). So, there's probably more to solve. I am merely hoping someone somewhere provides a native m1, m2, etc Apple Silicon builds for qownnotes. How are yours built for Intel and can't they be posted for Apple Silicon too?

**Comment by pbek** [comment]
> How are yours built for Intel and can't they be posted for Apple Silicon too?

The whole build chain, release process and update mechanism is structured around one single macOS package...

> While the build works, the main toolbar is messed up in that half the icons are missing 

Hm, can you turn on the internal icon theme?

> I don't care if nix works per se, what I am looking for is simply a native Mac build (on Apple Silicon not Intel) of qownnotes

If there is no other way, then the nix release could be that build. Nix can also update that package.

**Comment by pbek** [comment]
The nixos package from https://github.com/pbek/QOwnNotes/blob/main/default.nix should build now.
(See https://github.com/pbek/QOwnNotes/actions/runs/12243377462).

But I wonder which platform the regular macOS app build is current running on... It does look as if `macos-latest` is already on aarch64. 🤔

@sfatula, @KwisatzJim could you maybe send the first paragraph from the debug settings of the regular build?

**Comment by sfatula** [comment]
If I get a chance before flying out I can try it. Meanwhile, I turned on the internal theme (restarted of course), same issue. So, I turned off dark mode icon theme, still no change. 

Here's my debug settings from the regular intel build:

`## General Info

**Current Date**: `Mon Dec 9 22:02:22 2024`
**Version**: `24.9.2`
**Build date**: `Sep  5 2024`
**Build number**: `1124`
**Platform**: `macosx (cocoa)`
**Operating System**: `macOS 14.7`
**Build architecture**: `x86_64`
**Current architecture**: `x86_64`
**Release**: `GitHub Actions`
**Qt Version (build)**: `5.15.2`
**Qt Version (runtime)**: `5.15.2`
**Portable mode**: `no`
**Settings path / key**: `/Users/sfatula/Library/Preferences/com.pbe.QOwnNotes.plist`
**Application database path**: `/Users/sfatula/Library/Application Support/PBE/QOwnNotes/QOwnNotes.sqlite`
**Application arguments**: `/Applications/QOwnNotes.app/Contents/MacOS/QOwnNotes`
**Qt Debug**: `0`
**Locale (system)**: `en_US`
**Locale (interface)**: *empty*
**Primary screen resolution**: `2560x1440`
**Icon theme**: `breeze-dark-qownnotes`
**Notes in current note folder**: `220`
**Calendar items**: `0`
**Enabled scripts**: `1`
**Database drivers**: `QSQLITE, QODBC, QODBC3, QPSQL, QPSQL7``

**Comment by sfatula** [comment]
For the one I built somewhat and mostly works but has (at least) the icon issues, it looks like this on Apple Silicon:

`## General Info

**Current Date**: `Tue Dec 10 00:26:43 2024`
**Version**: `24.11.1`
**Build date**: `Nov  8 2024`
**Build number**: `1138`
**Platform**: `macosx (cocoa)`
**Operating System**: `macOS Sonoma (14.7)`
**Build architecture**: `arm64`
**Current architecture**: `arm64`
**Release**: `GitHub`
**Qt Version (build)**: `6.8.0`
**Qt Version (runtime)**: `6.8.0`
**Portable mode**: `no`
**Settings path / key**: `/Users/sfatula/Library/Preferences/com.pbe.QOwnNotes.plist`
**Application database path**: `/Users/sfatula/Library/Application Support/PBE/QOwnNotes/QOwnNotes.sqlite`
**Application arguments**: `./QOwnNotes`
**Qt Debug**: `0`
**Locale (system)**: `en_US`
**Locale (interface)**: *empty*
**Primary screen resolution**: `2560x1440`
**Icon theme**: `breeze-dark-qownnotes`
**Notes in current note folder**: `220`
**Calendar items**: `0`
**Enabled scripts**: `1`
**Database drivers**: `QSQLITE, QODBC, QPSQL``

**Comment by pbek** [comment]
Thank you.
This version should also give you shell completion: https://github.com/NixOS/nixpkgs/pull/363327/files

**Comment by pbek** [comment]
## 24.12.2
- The nix package now builds also for macOS on x86_64 and aarch64, aka. Apple Silicon
  (for [#2912](https://github.com/pbek/QOwnNotes/issues/2912))
  - There is an open pull request for it at [#363327 on NixOS/nixpkgs](https://github.com/NixOS/nixpkgs/pull/363327)
  - This could serve as a native package for macOS on Apple Silicon hardware

**Comment by sfatula** [comment]
So, I did use that default.nix file, it does run to completion. I do not find a .app file, I searched everything under /nix/store. And it's not in /Applications/Nix Apps either. A find command finds it nowhere. I do see the binary file again (in my case at /nix/store/d929s2zwg8lybx9yd3g504w359c4820f-qownnotes-24.11.2/bin/) but that's not a Mac app. It can be run but it should be packaged like all other Mac apps with a .app extension and a bunch of stuff underneath that. I would guess most any Mac app on nix has a packager to do so. An example might be the commonLinux program remmina which I use.

The other good news is the icons all look good.

**Comment by pbek** [comment]
>  I do not find a .app file

If you compile something that is not part of your nix configuration, nix creates a `./result` folder.

> The other good news is the icons all look good.

The icons from which build?

**Comment by sfatula** [comment]
I made it part of my nix configuration as I didn't know how to compile something outside of my nix configuration, no ./result folder. I overlaid the nix default.nix with this one and did a darwin-rebuild. So, just as if someone used nix to build it, no .app (Mac app) file, just the binary that gets included amongst other things in the .app file (which is actually a folder). I only had one app being built with nix, remmina as it was the only source for Apple Silicon version. So, not much need to learn anything about it. 

The icons from the new build work fine. From previous build as I had noted above a few days ago, it did not. I can turn the binary into a app "file" via automator or other simple method. I just htink it should build a proper .app "file" so others than may use it don't have to mnss around.

**Comment by pbek** [comment]
> The icons from the new build work fine.

The aarch64_darwin nix build or the x86 macOS release build?

> I just htink it should build a proper .app "file" so others than may use it don't have to mnss around.

I've found some hints of packages dealing with an app file, but I'm not sure yet why it wasn't created in the build process.

https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22.app%22+path%3A*.nix&type=code

In the release build, I create the app file in the build script with macOS tooling...

**Comment by sfatula** [comment]
The aarch64_darwin build, icons work perfectly (as did the original x86 build). It did not work with the first nix build made originally with your initial nix changes.  I don't have a x86 machine. Mac with m# chip can run x86 binaries but it's via a code translater and you have to install it. Thus far, everything works with the arm build. Been using it for a day. How nix works is beyond me, never looked at the developer doc. But you asked them so hopefully they have an answer or maybe it just works once included in their entire system.

**Comment by pbek** [comment]
https://github.com/NixOS/nixpkgs/pull/363327 is now merged into `master` and will be soon in nixos-unstable. You can track the progress here: https://nixpk.gs/pr-tracker.html?pr=363327

**Comment by sfatula** [comment]
Just an FYI. On the latest stable nix (25.05), qownnotes 25.5.3 works, however, there is an error when quitting qownnotes. Here it is:

```
sfatula@Mac:/run/current-system/sw/bin/ > ./qownnotes
Warning: Unable to find dictionary for "en_US" in path "/en_US" (:0, )
Warning: QThreadStorage: entry 2 destroyed before end of thread 0x600002358070 ((null):0, (null))
zsh: segmentation fault  ./qownnotes
```

**Comment by pbek** [comment]
Thank you. It basically crashes. 😆

**Comment by sfatula** [comment]
It works for days or weeks, only when you quit the app (or reboot) does it crash. If there is anything I can do to help debug it, let me know.

**Comment by pbek** [comment]
I can try adding something blindly and ask you if it helps in the next release.

**Comment by sfatula** [comment]
We'll see how that works once it comes out.

**Comment by pbek** [comment]
@sfatula, the thing I wanted to try didn't work. Could you maybe open a separate issue and show your settings and debug log output?

**Comment by sfatula** [comment]
Yes, I will do so once I get a chance. Been repairing things around the house for 2 weeks, house got struck by lightning and so many things broken.

**Comment by pbek** [comment]
Oh my god, that sounds like a nightmare!

**Comment by sfatula** [comment]
Have been very lucky. Heat pump broke, they did diagnostics and blower motor needed replaced. And a couple days later before it came in, magically started working and has since. Home theater receiver broke, but they covered that under warranty. Garage door opener broke, but they covered that under warranty. The rest are smaller devices like routers, wall switches, irrigation, etc. which I've been replacing. So, nowhere near a homeowner insurance claim fortunately, just a ton of time. For a while, I'd fix something only to discover 2 more things now are broken. Sorry for off topic!

**Comment by pbek** [comment]
The cables and pipes are intact? Yes, you are almost lucky 😅

**Comment by jefftemplon** [comment]
Hi, I'm curious about the progress on non-Nix (homebrew or App Image) builds with Apple Silicon support.  I'm ready to give this product a serious try, the lack of native builds is the only blocker.

**Comment by pbek** [comment]
No news. If someone finds a way to build a universal binary in the GitHub CI, I'm all ears...

**Comment by jefftemplon** [comment]
I'm a GitHub CI idiot, I'm no help there.  I did find this link, which may help.

https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary

**Comment by jefftemplon** [comment]
There's also this: https://github.com/marketplace/actions/macos-universal-binary-action

**Comment by pbek** [comment]
> There's also this: https://github.com/marketplace/actions/macos-universal-binary-action

Looks interesting, but I have no idea how to use the action with a custom build process... ️🤷🏻

**Comment by jefftemplon** [comment]
I just built the Apple silicon binary from source, I was rather impressed that it "just worked".  How hard is it to have two separate binaries for Mac: one intel and one Apple silicon?  Maybe that is easier than making a universal binary.

**Comment by pbek** [comment]
The (hand-crafted) update mechanism only works for one.

**Comment by sfatula** [comment]
Consider that Intel Macs will not be supported for long. The next MacOS only supports 4 models as far as I understand it, and, even the Rosetta translation from Intel apps to run on Applie Silicon will be going away. Given that, at some point you should consider which is more important going into the future, M# or old Intel. There would be other ways to get the app for Intel so, maybe Apple Silicon more important now or soon?

**Comment by pbek** [comment]
Hm, too bad I can't see how many are on Intel still (because Rosetta is emulating an Intel machine), I still know people with Intel Macs... And those machines are still good. ️🤷🏻 At some point, I'll try to use the arm runners... Until then, QOwnNotes seems to work fine via nix.

**Comment by jefftemplon** [comment]
Thanks for your clarity!

**Comment by pbek** [comment]
I tried to build on GitHub with x86_64 and arm64 support, I'm not sure what came out of it. Does anyone care to test https://github.com/pbek/QOwnNotes/actions/runs/16176553627/artifacts/3497945665?

**Comment by sfatula** [comment]
It's definitely a universal build. It's running as native, not via Rosetta. Version says 25.7.5

A quick look says this is great! Would have to use for a while obviously. Good job!

**Comment by pbek** [comment]
@sfatula, can you please install the script `GitHub headlines` from the script repository, copy `https://github.com/pbek/QOwnNotes/issues/2912` and paste it with `Cmd + Shift + V` to a note to test scripting and HTTPS? And do you have Nextcloud integration on by chance?

**Comment by sfatula** [comment]
I am using the nextcloud integration, yes. Versioning does work. I will test the other when I get a chance, have had no AC for 10 days and working with the heat/air guys.

**Comment by pbek** [comment]
Thank you! Scripting would be the last step for me, considering going straight for the Qt 6.10 universal binary build for the main release!

**Comment by sfatula** [comment]
Didn't scripts used to show up in Scripting -> Custom Actions in the menu bar? I don't see any there, neither my Fix LIne Numbers or the newly installed Github Headlines. 

I will note when running the new app you just built, it prompts over and over (around 10 times) when starting the app about needing removable volume permission. Is the app not signed perhaps?

**Comment by pbek** [comment]
> Didn't scripts used to show up in Scripting -> Custom Actions in the menu bar?

Only if they have a custom action.
Is there any useful log output if you enable the log panel (see Window / Panels)? Please make sure you also enable debug logging in the log panel (by right-clicking on the panel and selecting Debug in Options).

> Is the app not signed perhaps?

No, it's not signed.

**Comment by sfatula** [comment]
I can fix the permissions issues with this:

`lipo QOwnNotes.app/Contents/MacOS/QOwnNotes -extract arm64 -output QOwnNotes.app/Contents/MacOS/QOwnNotes`

This removes the x86 version from the universal binary and leaves only the arm version. Apparently this started in qt6? 

I will get to the debug later.

**Comment by sfatula** [comment]
I have the log panel up with debug logging and not sure it's helpful as there simple is nothing under custom actions, so, nothing to debug maybe. 

I quit and restarted to see if anything shows up, I see this:

`[18:23:06] [debug] updatePanelMenu - 'updatePanelMenu'
[18:23:06] [debug] start searchForUpdates
[18:23:06] [debug] on_noteTreeWidget_itemSelectionChanged
[18:23:06] [debug] setCurrentNote - 'note': Note: <id>14 <name>"UTM" <fileName>"UTM.md" <noteSubFolderId>1 <relativePath>"Mac" <hasDirtyData>false - 'updateNoteText': true - 'updateSelectedNote': false
[18:23:06] [debug] added to history: NoteHistoryItem: <noteName>"UTM" <noteSubFolderPathData>"Mac" <cursorPosition>99 <relativeScrollBarPosition>0
[18:23:07] [debug] Reply from "/index.php/apps/qownnotesapi/api/v1/note/app_info"
[18:23:07] [debug] Reply from app version test
[18:23:07] [debug] Reply from "/ocs/v2.php/apps/files_sharing/api/v1/shares"
[18:23:07] [debug] Reply from share api
[18:23:07] [debug] updateNoteShareStatusFromShare - 'data': "<?xml version=\"1.0\"?>\n<ocs>\n <meta>\n <status>ok</status>\n <statuscode>200</statuscode>\n <message>OK</message>\n </meta>\n <data/>\n</ocs>\n"
[18:23:07] [debug] updateNoteShareStatusFromShare - 'status': "ok"
[18:23:07] [debug] updateNoteShareStatus - 'dataElements.count()': 1
[18:23:07] [debug] updateNoteShareStatus - 'shareType': ""
[18:23:07] [debug] updateNoteShareStatusFromFetchAll - 'data': "<?xml version=\"1.0\"?>\n<ocs>\n <meta>\n <status>ok</status>\n <statuscode>200</statuscode>\n <message>OK</message>\n </meta>\n <data/>\n</ocs>\n"
[18:23:07] [debug] updateNoteShareStatus - 'dataElements.count()': 1
[18:23:07] [debug] updateNoteShareStatus - 'shareType': ""
[18:23:07] [debug] "Found scripts: 105"
[18:23:07] [debug] Checking for script update: QUrl("https://raw.githubusercontent.com/qownnotes/scripts/master/fix-list-numbers/fix-list-numbers.qml")
[18:23:07] [debug] Checking for script update: QUrl("https://raw.githubusercontent.com/qownnotes/scripts/master/github-headline/github-headline.qml")
[18:23:07] [status] 🔧 No script updates were found
[18:23:07] [debug] done searchForUpdates`


So, it sees the 2 scripts, but, menu option not populated for some reason. I use fix lists to renumber them. Works on the nix version. Not sure why the code block not formatting it right.

Also, fyi, when debug logging is on, and, the panel is displayed, and I quit the app, I get the problem I had before you fixed it a while back, the app crashes on quit, but only with those settings on.

**Comment by pbek** [comment]
> This removes the x86 version from the universal binary and leaves only the arm version. 

I wonder how that removes the macOS errors/warnings. And I guess you would need to run that after each update of the application. Can you maybe post a screenshot of the warnings?

> Apparently this started in qt6?

Or maybe you don't get the errors in Rosetta? Did you get the errors in the nix version?

> I quit and restarted to see if anything shows up, I see this:

Can you please send a log of when you reload the scripting engine?

> Also, fyi, when debug logging is on, and, the panel is displayed, and I quit the app, I get the problem I had before you fixed it a while back, the app crashes on quit, but only with those settings on.

Hm, also annoying...

**Comment by sfatula** [comment]
No, that is from various bug reports, it is a feature of qt6 on Mac. Many qt6 apps have had the same issue. As far as I know, no non qt6 app has the issue (there probably some obscure one somewhere). 

The lipo command would indeed need done after each release. I personally don't care, no big deal to me. I suppose you could make a small installer script until qt6 is fixed. Or a note, whatever. Heck, you could send it to me and I'll send it back. But it appears it can run on Linux if I read this correctly:

https://github.com/konoui/lipo

I will send such a log tomorrow sometime, it's 2:40AM here. And it's 79 in the house, still have no AC.

**Comment by pbek** [comment]
> The lipo command would indeed need done after each release. I personally don't care, no big deal to me. I suppose you could make a small installer script until qt6 is fixed.

I don't think this will be "fixed", because Apple wants you to pay their tax and sign your Applications manually, which I will not do.

> Or a note, whatever. Heck, you could send it to me and I'll send it back.

Thank you, but that doesn't work either as part of the automatic build and release process and because I don't know what it will do on Intel systems. 😉

> I will send such a log tomorrow sometime, it's 2:40AM here. And it's 79 in the house, still have no AC.

Thank you!

**Comment by sfatula** [comment]
I have TONS of unsigned apps, and they all run. It's a qt6 thing, not an unsigned app thing. 

As far as Intel, the idea here is you'd have the universal binary for Intel, AND the arm version for the majority of Macs out there. So, you'd have both from one build. The intel would remain unchanged.

**Comment by pbek** [comment]
I see. Yes, it wasn't the case with Qt5.

**Comment by sfatula** [comment]
It actually might be a few days, I will get to it, today is warmer and I will not have access to my Mac, or tomorrow either as I have to keep that part of the house hot to keep the part we are in cooler. Waiting on parts!

**Comment by pbek** [comment]
I have no AC, I like it hot. 😁

**Comment by sfatula** [comment]
The heat index is 121 degrees outside. Dew point is 81, very humid. Trying to keep my equipment like routers, nas, etc from melting. Normally I'd be working in the yard but I can't come in to cool off. If it gets much worse, will have to abandon the house until the part(s) come in. It;s a dangerous situation esp for me and my health.

Another option since another user posted how easy it is to build is to put it in homebrew, vastly more popular than nix on Mac. It could build there.

**Comment by sfatula** [comment]
> I just built the Apple silicon binary from source, I was rather impressed that it "just worked". How hard is it to have two separate binaries for Mac: one intel and one Apple silicon? Maybe that is easier than making a universal binary.

So, what build instructions did you follow? Maybe I can build it and then get it into Homebrew.

**Comment by jefftemplon** [comment]
I can't find the link anymore, but it's basically the same as the documentation, only before following that you need to do 
```
brew install qt-creator
brew install qt
```
which will pull in a bunch of dependencies; then follow the build instructions on the website.

**Comment by pbek** [comment]
> The heat index is 121 degrees outside.

😲 Ok, my comfy-temp is at 30°C... 50°C is insane

> Doc

https://www.qownnotes.org/installation/macos.html
or
https://www.qownnotes.org/installation/building.html

**Comment by sfatula** [comment]
[11:39:49] [scripting] loading script file: /Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/fix-list-numbers/fix-list-numbers.qml
[11:39:49] [warning] QQmlComponent: Component is not ready
[11:39:49] [warning] script errors: QList(file:///Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/fix-list-numbers/fix-list-numbers.qml:1:1: module "QtQml" is not installed import QtQml 2.0 ^)
[11:39:49] [scripting] loading script file: /Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/github-headline/github-headline.qml
[11:39:49] [warning] QQmlComponent: Component is not ready
[11:39:49] [warning] script errors: QList(file:///Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/github-headline/github-headline.qml:1:1: module "QtQml" is not installed import QtQml 2.0 ^)
[11:39:49] [status] 🔧 The scripting engine was reloaded

**Comment by pbek** [comment]
>  [11:39:49] [warning] QQmlComponent: Component is not ready

That means that macdeployqt doesn't deploy the QML libraries needed for scripting... 🙄

**Comment by sfatula** [comment]
I should point out that I don't think it will ask for removable volume permissions on most Macs. Likely the only reason it does for me is my nextcloud notes folder is on an external drive. 

That can be resolved by the user doing this (after each download, it's a workaround) without doing the lipo (i.e. running your test app you built, the universal image):

sudo codesign -s - --deep /Applications/QOwnNotes.app

And only needed for those running their notes folder on an external drive. So, probably not so bad. The remaining issue other than the crash below is scripting not working. Otherwise, seems to work fine! 

Did a little more testing. If the log panel is displayed, it crashes when quitting the app. Debug on or not does not matter.

**Comment by pbek** [comment]
Ok, thank you. I noted the command down. But still, scripting is a major issue. Haven't found a way to get the QML libraries deployed. The logging panel crash is annoying too, but that can only be fixed by someone with an Apple Silicon machine, which I have not...

**Comment by sfatula** [comment]
Last time didn't you add some sort of debugging to figure out where? Or was it I sent you a crashlog? I don't recall really, still hot!

[Crashlog.txt](https://github.com/user-attachments/files/21202741/Crashlog.txt)

**Comment by pbek** [comment]
Thank you for the crashlog.
I can try another blind attempt to fix it.

**Comment by pbek** [comment]
## 25.7.6

- Try to fix a crash on macOS Apple Silicon on Qt6 while quitting the app with
  log panel turned on (for [#2912](https://github.com/pbek/QOwnNotes/issues/2912))

The build should be available here soon:
https://github.com/pbek/QOwnNotes/actions/runs/16273426645

**Comment by sfatula** [comment]
Still crashed

[Crashlog.txt](https://github.com/user-attachments/files/21220843/Crashlog.txt)

**Comment by pbek** [comment]
Thank you, I wonder what is macOS doing differently than all the other platforms...

**Comment by sfatula** [comment]
Not sure, but this may drive you more nuts. I built qownnotes from source using qt creator and qt6, and, script shows up and qownnotes quits just fine with log panel on. 

Here's the log window for reloading scripting:
[00:42:03] [scripting] loading script file: /Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/fix-list-numbers/fix-list-numbers.qml
[00:42:03] [scripting] loading script file: /Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/github-headline/github-headline.qml
[00:42:03] [status] 🔧 The scripting engine was reloaded

And, the permissions problem also does not show up. Therefore, I suspect the build process you are using. It works perfectly. 

QOwnNotes - version 25.7.6 - Self-build - build 1203
Plain-text file markdown note taking with Nextcloud/ownCloud integration for GNU/Linux, macOS® and Windows® (it is written in C++/Qt®), that works together with the notes application of [Nextcloud](https://nextcloud.com/) and [ownCloud](https://owncloud.org/). 
Built on Jul 15 2025 with Qt 6.9.1.

**Comment by pbek** [comment]
Yes, if you build from source (or the nix package is used) the QML libraries are present. But the libraries are not present in the DMG. It's a deploying issue. You can't tell `macdeployqt` to deploy additional libraries.  [Qt for macOS - Deployment | Qt 6.9](https://doc.qt.io/qt-6/macos-deployment.html)

**Comment by sfatula** [comment]
The application bundle I built has the frameworks in it, such as QtQml. You are saying macdeployqt doesn't put them in the bundle? But did in qt5?

The doc you linked to says it will put them there. It's kind of useless without it. Have you run it with verbose=3? Are you using the latest version of macdeployqt?

**Comment by pbek** [comment]
> The application bundle I built has the frameworks in it, such as QtQml. You are saying macdeployqt doesn't put them in the bundle? But did in qt5?

Now that you mention it, I can see them in https://github.com/pbek/QOwnNotes/actions/runs/16301194939/job/46035974296#step:6:3670.

Could you maybe compare the libraries from the Qt5 release and the Qt6 build? I can't open the DMG under Linux.

> 
> The doc you linked to says it will put them there. It's kind of useless without it. Have you run it with verbose=3? Are you using the latest version of macdeployqt?

I'm trying `-verbose=3` now...
For Ubuntu for example, I had to add three packages additionally with Qt6: https://github.com/pbek/QOwnNotes/blob/e20eb830ed82916e3827950118f5e9dc58e7e384/ubuntu-launchpad/qt6/control#L40-L42

**Comment by sfatula** [comment]
I don't have a qt5 version, can I get it somewhere? Or, I could email or send you the extracted dmg

**Comment by pbek** [comment]
All macOS releases are built with Qt5: https://github.com/pbek/QOwnNotes/releases
I'm just interested in a list of files. Or better, the difference of files between the Qt5 and Qt6 version.

**Comment by sfatula** [comment]
Here is a list of files in the bundle that exists in one but not the other. For some reason, I thought you had converted to qt6 in that special new build you made above, just confused but not relevant, got your list. 

[QOwnNotes File DIfferences.txt](https://github.com/user-attachments/files/21241247/QOwnNotes.File.DIfferences.txt)

**Comment by pbek** [comment]
> For some reason, I thought you had converted to qt6 in that special new build you made above

No, I built a Qt6 release and store the DMG as build artifact.

Thank you for the diff, I can't see anything significant different. And yet, scripting is not working!? 🤔
Maybe things like https://doc.qt.io/qt-6/qml-qtqml-workerscript-workerscript.html and https://doc.qt.io/qt-6/qtqmlmodels-index.html are in a different library...

**Comment by sfatula** [comment]
I don't know enough about QT. So, the version I am saying scripting does not work (and the quit issue, and the security issue) in is your test builds above:

https://github.com/pbek/QOwnNotes/issues/2912#issuecomment-3070367323 and
https://github.com/pbek/QOwnNotes/issues/2912#issuecomment-3053632956

I have not tested AFAIK the existing version releases which wouldn't be for ARM. Just to be clear.

**Comment by pbek** [comment]
All macOS builds in this workflow are done with Qt6 as universal binary as build artifact: https://github.com/pbek/QOwnNotes/actions/workflows/build-release-qt6.yml

The regular release for macOS, that is promoted by the documentation and used by the qownnotes api and the internal updater, is still Qt5 intel.

**Comment by sfatula** [comment]
So, wouldn't you really want to compare the test version you built (on the qt6 workflow?) with the self built version I am running which all works and see what is different? Did verbose tell you anything?

**Comment by pbek** [comment]
Did you use macdeployqt to bundle an app file?
Verbose was too verbose to tell anything. What I need is an option to bunlde extra libraries.

**Comment by sfatula** [comment]
I built from qt-creator as qmake gave me an immediate error about qtwebsockets missing, which it isn't AFAIK. Whatever qt-creator does. The qt documentation page you had noted above somewhere mentioned a command line tool to add more stuff to the bundle. The page that had macdeployqt in it, it was a different command. I am heading out so can't look right now.

**Comment by pbek** [comment]
Maybe you need to install qtwebsockets in the Qt installer by hand...

That's how it's built in CI: https://github.com/pbek/QOwnNotes/blob/cceab5f3160f5ccaae74c8ed560ff261b332e536/.github/workflows/build-release-qt6.yml#L384-L386

And then the release shell script:
https://github.com/pbek/QOwnNotes/blob/054b76eeac6088fdfeb098e87ea8775543b89477/build-systems/github/macos/create-dmg-qt6.sh

**Comment by sfatula** [comment]
But I don't want to add it, using qt-creator to build qt apps is trivial, I used it and it works first try. I have a working build of QOwnNotes. Homebrew has to include qtwebsockets else I couldn't make my build even with qt-creator, so, the question is how do I point to it? I see it in /opt/homebrew/lib

You are wanting me to use your technique for some reason?

In the doc you referenced earlier for Mac qt deploment tool, there is a "install_name_tool" and that section seems to show how to copy in new frameworks.

**Comment by pbek** [comment]
> But I don't want to add it

Add what? QtWebsocket? It's a dependency of QOwnNotes. I can't remember if the checkbox in the Qt installer is checked by default. Or didn't you install QtCreator "by hand"? Is it in homebrew (couldn't find it)? All Qt stuff is a lot of clutter 😊

> You are wanting me to use your technique for some reason?

The main part is to build an App file with all libraries, translation files and Qt dependencies (macdeployqt) that can then be bundled in a DMG file.

**Comment by sfatula** [comment]
I changed my post. I am not using the qt installer but homebrew which compiles qt6 natively for Apple Silicon. I can find the qt libraries, so, how do I make qmake see them? See my re-written post

**Comment by pbek** [comment]
> install_name_tool

We don't know yet what's missing. For Ubuntu it was:
https://github.com/pbek/QOwnNotes/blob/e20eb830ed82916e3827950118f5e9dc58e7e384/ubuntu-launchpad/qt6/control#L40-L42

**Comment by pbek** [comment]
> I changed my post. I am not using the qt installer but homebrew which compiles qt6 natively for Apple Silicon. I can find the qt libraries, so, how do I make qmake see them? See my re-written post

I don't think it's about qmake, as it is just building the binary.
Or are you talking about qwebsocket? that needs to be installed or qmake can't find it.

**Comment by sfatula** [comment]
Ok, but I tcan't use qmake, it gives me:

`qmake QOwnNotes.pro CONFIG+=release
Project MESSAGE: Currently in RELEASE mode.
Project WARNING: Qt has only been tested with version 12 of the platform SDK, you're using 15.
Project WARNING: This is an unsupported configuration. You may experience build issues, and by using
Project WARNING: the 15.5 SDK you are opting in to new features that Qt has not been prepared for.
Project WARNING: Please downgrade the SDK you use to build your app to version 12, or configure
Project WARNING: with CONFIG+=sdk_no_version_check when running qmake to silence this warning.
Project ERROR: Unknown module(s) in QT: websocket`

But, there is a qtwebsockets as part of homebrew dir I noted (/opt/homebrew/lib). No clue how to make that work.

**Comment by pbek** [comment]
For Windows it was:

```ps
windeployqt -core -gui -widgets -sql -svg -network -xml -printsupport -qml -websockets -concurrent QOwnNotes.exe
```

I added `concurrent` so it will be copied to the output directory.

**Comment by pbek** [comment]
> But, there is a qtwebsockets as part of homebrew dir I noted (/opt/homebrew/lib). No clue how to make that work.

Where does your QT_DIR env var point to?

**Comment by pbek** [comment]
Btw., you don't need to use qmake, you can also use cmake (as it is done in the nix package). I just never did that manually under macOS.
My end goal is it to get every build system to Qt6 and cmake anyway.

**Comment by pbek** [comment]
Btw. in the GitHub CI, I use https://github.com/miurahr/aqtinstall to install qt.

**Comment by sfatula** [comment]
Ok, so, I had a second qmake and it was the issue. I found another qt install too, I don't recall why I have that. So, got to clean some stuff up.

**Comment by sfatula** [comment]
So, here's what I don't get then. I did this:

git clone https://github.com/pbek/QOwnNotes.git -b release --depth=1
cd QOwnNotes
git submodule update --init
cd src
/opt/homebrew/opt/qt/bin/qmake QOwnNotes.pro CONFIG+=release
make -j4

And the result in the src directory is a QOwnNotes.app for Apple Silicon that runs perfectly as is. It's a small application bundle, no frameworks in it at all . I presume then it links to stuff installed on my machine already, so, it doesn't need frameworks and such. How do I compile as if it was for someone else though without qt installed?

**Comment by pbek** [comment]
> And the result in the src directory is a QOwnNotes.app for Apple Silicon that runs perfectly as is.

Yeah, because you have installed all the libraries yourself and the app finds them. 😊

But that doesn't help us with the deployment of a DMG files with an APP file (directory) containing the needed frameworks/libraries, so that everyone else can also run the application. 😅 That's what macdeployqt is doing.

**Comment by sfatula** [comment]
I've never used that tool. So, if I use it, then, the result should be not be built for me right? And therefore, I could play with it until it works? 

I could start with your version of the command and see if it also fails first, and if so, fix it?

If that is the case, I can make it work.

**Comment by pbek** [comment]
As long there is an ".app" file in the end with a universal binary and all the Qt libraries, we are happy. 😊

**Comment by pbek** [comment]
@sfatula, I tried to add the QtConcurrent framework manually, could you please try https://github.com/pbek/QOwnNotes/actions/runs/16362525727/artifacts/3560709540 to see if scripts are working?

**Comment by sfatula** [comment]
Ok, only had to say yes to removable volumes permissions 63 times! I hope to solve that also for us external drive users. 

[00:22:01] [debug] filterNotesByNoteSubFolders - 'noteSubFolderIds': QList(3)
[00:22:14] [scripting] loading script file: /Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/fix-list-numbers/fix-list-numbers.qml
[00:22:14] [warning] QQmlComponent: Component is not ready
[00:22:14] [warning] script errors: QList(file:///Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/fix-list-numbers/fix-list-numbers.qml:1:1: module "QtQml" is not installed import QtQml 2.0 ^)
[00:22:14] [scripting] loading script file: /Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/github-headline/github-headline.qml
[00:22:14] [warning] QQmlComponent: Component is not ready
[00:22:14] [warning] script errors: QList(file:///Users/sfatula/Library/Application Support/PBE/QOwnNotes/scripts/github-headline/github-headline.qml:1:1: module "QtQml" is not installed import QtQml 2.0 ^)
[00:22:14] [status] 🔧 The scripting engine was reloaded

Isn't the key thing: module "QtQml" is not installed.

If you give me some time, I will try and find a way to resolve all the issues. It might save you some time. Unless you just can't stop, like me sometimes.

**Comment by pbek** [comment]
Thank you for testing and feedback!

Could you please do me a favor and open two new separate issues for the permissions and the scripting? And please mention macos and Qt 6.

**Comment by pbek** [comment]
> Isn't the key thing: module "QtQml" is not installed.

But it always was 🙄🤔
And I couldn't find different libraries for worker scripts, like under Ubuntu.

**Comment by pbek** [comment]
And thanks a lot for your help!

**Comment by sfatula** [comment]
Ok, I've resolved at least when building on my Mac two issues.

I removed the commented line in QOwnNotes.pro that was:
#QT       += quick

I removed the HASH, i.e., seems to be different on qt6, that fixed the script issue, qt quick was missing from the bundle apparently, and that added in turn a few other things.

For the endless permission issue, I used this macdeployqt line:

macdeployqt QOwnNotes.app -codesign=-

Now, I have only built for Apple Silicon, so, whether or not it works for Intel builds also, don't know.

So, I can run the deployed version with the copied frameworks and it works. I do not note any error when quitting the app either. 

I'm note sure of the codesign impact when sent to another person though, so, may or may not be better, you should build and send to me. 

I guess I should have mentioned that before I made any of the above adjustments, I ran the macdeployqt version using my build and it had the same errors as your build. So, my changes definitely did something.

**Comment by pbek** [comment]
Now that's great news.
I'm trying to build in https://github.com/pbek/QOwnNotes/actions/runs/16408836666, does anyone have an intel mac to test the artifact?

**Comment by sfatula** [comment]
We'll see if it's great news! I may or may not be able to test it tonight. I am about to retire to bed. We'll see. This week I am very busy with stuff every evening, so, hopefully, it just works!

**Comment by pbek** [comment]
@sfatula (and anyone else with intel or Apple Silicon), could you please test https://github.com/pbek/QOwnNotes/actions/runs/16408836666/artifacts/3574665998 for warnings, scripting and the crash on exit when logging panel is turned on?

**Comment by LockeBirdsey** [comment]
> ...could you please test https://github.com/pbek/QOwnNotes/actions/runs/16408836666/artifacts/3574665998 for warnings, scripting and the crash on exit when logging panel is turned on?

Tested that build on a basically fresh Silicon Mac and a much less-fresh Intel Mac. It looks like everything is behaving properly. The logging panel crash happens on both, but that bug seems to be plaguing lots of Qt6 projects (e.g. OBS). And more interestingly the Intel Mac will always "crash" on quit now, regardless of which panels are open.

**Comment by sfatula** [comment]
Works perfectly on Silicon, normal first time security permission approval and scripts work. No crash. So, sounds like Intel has a slight issue with quitting according to the other poster.  I have read about the crash problem with qt6. Not sure if there is a solution or not, but also not sure if your previous qt6 builds had the issue either.

**Comment by pbek** [comment]
@LockeBirdsey, thank you for testing. Did you also test if scripts are working?
@sfatula, thank you for testing.

**Comment by LockeBirdsey** [comment]
> @LockeBirdsey, thank you for testing. Did you also test if scripts are working?

Yep, tested a few on both systems

**Comment by sfatula** [comment]
It took me a few hours to get it going as Homebrew version of qt6 has a problem with macdeloyqt. I ended up removing homebrew qt6, and installing from the qt online installer instead. That worked fine.  Was not expecting that so confused for a while. 

I will amend my earlier point. If the panel is displayed when quitting on Silicon, it does in fact have the quit error. Otherwise, it does not. It does not do that in the non universal version that I build. Or your version if I remove the Intel part. 

I think you could avoid it in the deploy script of yours if you keep separate intel and silicon builds. You could build the universal, then extract from the one build 2 different app files, one for each architecture. Not sure how that works in the environment you are using though.

**Comment by pbek** [comment]
> I think you could avoid it in the deploy script of yours if you keep separate intel and silicon builds. 

The updater script uses one disk image...

**Comment by sfatula** [comment]
I personally think it's acceptable, eventually the qt6 (suspected) or MacOS bug (possible) will be resolved. If you want a Silicon build, and, an Intel build, then there is no issue for the majority, Mac Silicon users, and there's simply a quit issue for Intel. Honestly, I don't ever quit the app anyway, just at shutdown. And I doubt people keep the log panel on permanently for the most part. (I know Intel has the problem panel or not). 

Not sure if you can fix it or not. Might poke around the internet looking for solutions from other qt6 projects if it's that troublesome. 

I personally would rather see the Universal build than the Intel build. Or, I could simply make my own. I have to suspect someone has found a workaround though.

**Comment by pbek** [comment]
I was trying to get a macOS VM running with quickemu, but wasn't able to get a successful installation out of it yet... So I guess I need to work blindly...
I'd do a release with the universal binary soon.

**Comment by pbek** [comment]
@sfatula, could you please open a separate issues for the seg fault again, like #3290?

**Comment by sfatula** [comment]
I will, just busy but somewhat soon. Surprised quickemu didn't work. 

Here's another choice for Intel:

https://github.com/sickcodes/Docker-OSX?tab=readme-ov-file

**Comment by pbek** [comment]
> I will, just busy but somewhat soon.

Thank you!

> Surprised quickemu didn't work.

The installation of macOS in the VM run through, but it will not boot afterward. I just get back to the macOS boot menu.

> 
> Here's another choice for Intel:
> 
> https://github.com/sickcodes/Docker-OSX?tab=readme-ov-file

Thank you, could be an option. But I fear will not be working, because I don't have X11 anymore. 🤔

**Comment by pbek** [comment]
- After a huge amount of work over many weeks, there now is a **universal binary**
  for **macOS**, built with **Qt6** that works natively on **Intel and Apple Silicon Macs**
  (for [#2912](https://github.com/pbek/QOwnNotes/issues/2912))
  - A big thank you to everyone involved
  - The universal binary is available in the [releases on GitHub](https://github.com/pbek/QOwnNotes/releases)
    as `QOwnNotes.dmg` and is replacing the previous Intel Qt5 build
  - You need at least macOS 13 (Ventura) to run the new universal binary,
    the Qt5 legacy version will still be available as `QOwnNotesQt5.dmg`
  - A known issue with Qt6 on Intel Macs is that the application might crash on
    exit when the logging panel was turned on
  - Please visit the [macOS installation page](https://www.qownnotes.org/installation/macos.html)
    for more information

There now is a new release, could you please test it and report if it works for you?

**Comment by sfatula** [comment]
Working for me on Apple Silicon, no errors on exit even with log panel shown

**Comment by pbek** [comment]
Awesome!
